File: CPP/49.GroupAnagrams.cpp
#include <bits/stdc++.h>
#include <algorithm>
using namespace std;

template <typename T>
std::ostream& operator<<(std::ostream& os, const std::vector<T>& vec) {
    os << "[";
    if (!vec.empty()) {
        // Print the first element outside the loop to avoid a trailing comma
        os << vec[0];

        // Print the rest of the elements with a comma separator
        for (size_t i = 1; i < vec.size(); ++i) {
            os << ", " << vec[i];
        }
    }
    os << "]";
    return os;
}

// --------------------------------------------------------------- //
vector<vector<string>> groupAnagrams(vector<string>& strs) {
    vector<vector<string>> ans;
    map<string,vector<string>> m;
    for (string s: strs) {
        string sorted=s;
        sort(sorted.begin(),sorted.end());
        if (m.count(sorted)>0) { //is this needed???
            m[sorted].push_back(s);
        }
        else {
            m[sorted].push_back(s);
        }
    }
    for (auto [k, v] : m) {
        ans.push_back(v);
    }
    return ans;
}
int main() {
    vector<string> v1{"eat","tea","tan","ate","nat","bat"};
    cout << groupAnagrams(v1) << endl;

    

    return 0;
}

File: Rust/src/bin/49_group_anagrams.rs
use std::collections::HashMap;

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut ans_hm:HashMap<Vec<i32>, Vec<String>>=HashMap::new();
        
        for s in strs {
            let mut k=vec![0;26];
            for ch in s.chars() {
                let pos= ch as i32 -'a' as i32;
                k[pos as usize] +=1;
            }
            ans_hm.entry(k).or_insert(Vec::new()).push(s);
        }
        let ans=ans_hm.values().cloned().collect();
        ans
    }
}

struct Solution { }
fn main() {
    let strs1=["eat".to_string(),"tea".to_string(),"tan".to_string(),"ate".to_string(),"nat".to_string(),"bat".to_string()].to_vec();
    println!("{:?}",Solution::group_anagrams(strs1)); // [["bat"],["nat","tan"],["ate","eat","tea"]]

    let strs2=["".to_string()].to_vec();
    println!("{:?}",Solution::group_anagrams(strs2)); // [[""]]

    let strs3=["a".to_string()].to_vec();
    println!("{:?}",Solution::group_anagrams(strs3)); // [["a"]]
    
}